File: Library.java
import java.util.ArrayList;
public class Library {
    private ArrayList<User> users;
    private ArrayList<Book> books;
    private ArrayList<Book> notRentedBooks;
    private ArrayList<Book> rentedBooks;

    public Library(){
        this.users = new ArrayList<>();
        this.books = new ArrayList<>();
    }

    public ArrayList<User> getUsers() {
            return users;
        
    }

    public ArrayList<Book> getBooks() {
        return books;
    }

    public ArrayList<Book> getNotRentedBooks(){
        return notRentedBooks;
    }

    public ArrayList<Book> getRentedBooks(){
        return rentedBooks;
    }


    public void agregarUsuario(User user){
        users.add(user);
    }

    public void agregarLibro(Book book){
        books.add(book);
        
    }

    public void mostrarUsuarios(){
        for (User user : users) {
            System.out.println(user.getName());
        }
    }

    public void mostrarLibros(){
        for (Book book : books) {
            System.out.println(book.getTitle());
        }
    }

    public void mostrarLibrosNoRentados(){
        for (Book book : books) {
           if (book.isRented()== false) {
            System.out.println(book.getTitle());
           }
        }
    }

    
    public void mostrarLibrosRentados(){
        for (Book book : books) {
           if (book.isRented()== true) {
            System.out.println(book.getTitle());
           }
        }
    }

    public void mostrarUsuariosQueHanRentado(){
        for (User user : users) {
           if (user.isRentedABook()== true) {
            System.out.println(user.getName());
           }
        }
    }

    public void rentarLibro(String name, String title) {
        User user = null;
        Book book = null;
        boolean userFound = false;
        boolean bookFound = false;
    
        for (User u : users) {
            if (u.getName().equals(name)) {
                user = u;
                userFound = true;
                break;
            }
        }
    
       
        for (Book b : books) {
            if (b.getTitle().equals(title)) {
                book = b;
                bookFound = true;
                break;
            }
        }
    
        if (userFound && bookFound) {
            if (!book.isRented()) {
                book.setRented(true);
                user.setRentedABook(true);
                System.out.println("Libro rentado por el usuario " + name);
            } else {
                System.out.println("El libro ya está rentado.");
            }
        } else {
            System.out.println("El usuario o el libro no existen en la librería.");
        }
    }

}

File: Library.py
from User import User
from Book import Book
class Library:
    __users = []
    __books = []
    __notRentedBooks = []
    __rentedBooks = []
    
    
    def __init__(self):
        self.__users = []
        self.__books = []
        self.__notRentedBooks = []
        self.__rentedBooks = []
        
    def get_users(self):
        return self.__users
    
    def get_books(self):
        return self.__books
    
    def get_not_rented_books(self):
        return self.__notRentedBooks
    
    def get_rented_books(self):
        return self.__rentedBooks
    
    def agregar_usuario(self, user):
        self.__users.append(user)
        
    def agregar_libro(self, book):
        self.__books.append(book)
        
    def mostrar_usuarios(self):
        for user in self.__users:
            print(user.get_name()+"\n")
            
    def mostrar_libros(self):
        for book in self.__books:
            print(book.get_title()+"\n")
            
    def mostrar_libros_no_rentados(self):
        for book in self.__books:
            if book.is_rented==False:
                print(book.get_title()+"\n")
            
    def mostrar_libros_rentados(self):
        for book in self.__books:
            if book.is_rented==True:
                print(book.get_title()+"\n")
                
    def mostrar_usuarios_que_han_rentado(self):
        for user in self.__users:
            if user.is_rented_a_book==True:
                print(user.get_name()+"\n")
                
    def rentar_libro(self, name, title):
        user = None
        book = None
        user_found = False
        book_found = False
        
        for u in self.__users:
            if u.get_name() == name:
                user = u
                user_found = True
                break
            
        for b in self.__books:
            if b.get_title() == title:
                book = b
                book_found = True
                break
         
        if user_found and book_found:
            if not book.is_rented():
                    book.set_rented(True)
                    user.set_rented_a_book(True)
                    print("\nLibro rentado por el usuario "+ name+"\n")
            else:
                print("\nEl libro ya está rentado\n")  
        
        else:
            print("\nEl usuario o el libro no existen en la libreria\n")              